app/io/flow/play/util/RefreshingReferenceAsync.scala

```
@@ -0,0 +1,157 @@
+  * Upon successful completion, the data is refreshed with the retrieved data, otherwise the cache is not refreshed.
+  *
+  * The class will not initialize if the retrieval function fails the first `maxAttempts` times to avoid querying a
+  * cache that has never been initialized.
+  */
+trait RefreshingReferenceAsync[T] {
+
+  def logger: RollbarLogger
+
+  private[this] def log: RollbarLogger = logger.withKeyValue("class", getClass.getName)
+
+  /**
+    * The scheduler to use to schedule the [[retrieve]] function every [[reloadInterval]] period.
+    */
+  def scheduler: Scheduler
+
+  /**
+    * The context to use to execute the [[retrieve]] function every [[reloadInterval]] period.
+    */
+  def retrieveExecutionContext: ExecutionContext
+
+  /**
+    * Interval between refreshes
+    */
+  def reloadInterval: FiniteDuration
+
+  /**
+    * Function retuning the data to cache.
+    * It is called every [[reloadInterval]] period.
+    * If successful, the cache is refreshed with the latest data, otherwise the cache is not refreshed.
+    */
+  def retrieve: Future[T]
+
+  /**
+    * Maximum number of attempts to retrieve the data.
+    * If the [[retrieve]] function fails [[maxAttempts]] times in a row, the cache will not refresh and will retry
+    * to retrieve data after [[reloadInterval]].
+    *
+    * Default: 3
+    */
+  def maxAttempts: Int = 3
+
+  /**
+    * Forces the cache to refresh. If successful, the cache is refreshed with the latest data, otherwise the cache is not refreshed.
+    *
+    * @return the result of the refresh call, potentially a failed [[Future]]
+    */
+  def forceRefresh(): Future[T] = doRefresh()
+
+  def get: T = cache.get().value
+
+  // load blocking and fail if the retrieval is not successful after maxAttempts
+  private[this] val cache: AtomicReference[AsyncResult[T]] =
+    Try(Await.result(doLoadRetry(1, maxAttempts), 10.minutes)) match {
```

10 minutes seems like a very long timeout. Presumably this will be used to periodically refresh some piece of data that is fetched via a remote call? Would 1 second be a more sensibel timeout?

---

That is on startup, to avoid the cache to be ever empty. This happens once, before the server accepts requests.

---

I'm aware, still I don't think blocking service startup potentially for 10 minutes is a good idea. Better fail fast and restart/redeploy.

---

Makes sense, changing to lower. `1.second` seems too low as it may well occasionally take longer.
We may want to make it a parameter if it becomes a problem.

---

👍